### PR TITLE
Contract implementation

### PR DESCRIPTION
**Pull request by Delja**
# Contract.

This pr is a copy of #2747 with some modifications. It has a dependency with the #2779 and #2788.

Adding programming by contract (Design by contract) in Nit language. Contracts works with nit annotations. I decided to cut the contract pr in a smaller part to facilitate reviewing (some feature are not available on it, see section futur feature).

## Annotations

To define a new contract you need to use the corresponding annotation. For example it's possible to define a contract that x parameter must be strictly greater than 5. To do it would be necessary to define the contract in the following way `expects(x > 5)`. All expressions returning a boolean (comparison...) can be used as a condition.

Three annotations were added:
	
- `expects` to indicate the conditions need to the execution of the methods
- `ensures` to indicate the conditions of guarantee at the end of the execution of the methods
- `no_contract` to remove the inherited contract

This syntaxe is choose because in the next version of c++ (20) (see [P1369](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1369r0.pdf) for a overview) the contract will be define with the keywords expects and ensures. So I think it would be interesting to keep the same principle. But if all nit developper prefer to use `require` and `ensure` it's fine to change it.

## Method contract (ensures, expects)

For each method it's possible to define preconditions (`expects`) and  post-conditions (`ensures`). If the call of the method satisfies the prerequisites (`expects`) of the method, the caller may assume that the return conditions (`ensures`) will be satisfied.

The method contracts can access all the parameters of the method as well as the set of attributes/methods visible in the context of the method. i.e the set of parameters and the set of methods and attributes of the current class can be used (attributes declare locally in the method can not be used). For post-conditions (ensures) the `result` attribute has been added to perform a check on the return value of the method. Currently it's not possible to refer an old value in a ensures (exemple old(length) == length + 1).

### Building contract phase

A phase is executed to check all the methods. This check is done to know if:

- The method is annotated (redefined or not)

- The method is a redefinition of a method already having a contract (i.e a method that does not add any new conditions to the existing contract).

When a contract is detected the code is `extended` to add the verification features. Two methods are created. The first method is the contract face. Its goal is to provide the call of the contract verification and the call of the original method. With this strategy, the original code of the method is preserved. The second created method is for the contract verification. 

##### Expect:
```nit
class MyClass
	fun foo(x: Int)
	is
		expects(x > 0)
	do
		[...]
	end
end
```
Representation of the compiled class
```nit
class MyClass
	fun foo(x: Int)
	is
		expects(x > 0)
	do
		[...]
	end

	fun _contract_foo(x: Int)
	do
		_expects_foo(x)
		foo(x)
	end	

	fun _expects_foo(x: Int)
	do
		assert x > 0
	end
end
```
##### Ensure:
```nit
class MyClass
	fun foo(x: Int): Bool
	is
		ensures(result == true)
	do
		[...]
		return true
	end
end
```
Representation of the compiled class

```nit
class MyClass
	fun foo(x: Int): Bool
	is
		ensures(result == true)
	do
		[...]
		return result
	end

	fun _contract_foo(x: Intl): Bool
	do
		var result = foo(x)
		_ensures_foo(x, result)
		return result
	end

	fun _ensures_foo(x: Int, result: Bool)
	do
		assert result == true
	end
end
```

## Inheritance

Contracts support redefinition and adding condition. Note that when a contract is define in a parent class, it's no longer possible to remove this contract on all the classes that inherit or redefine them. They only need to be increased according to different subtyping rules.

All preconditions (expects) can be weakened. i.e it's possible to provide a new alternative to validate the contract. This corresponds to the use of a logical OR between the old and the new conditions.

All post-conditions (ensure) can be consolidate. i.e the new condition of the contract will provide a new guarantee to the user of the method. This rule can be translates into a logical AND between the old and the new conditions.

#### Expect

```nit
class SubMyClass
	super MyClass
	
	redef fun foo(x: Int)
	is
		expects(x > 0, x == 0)
	do
		[...]
	end
	
	redef fun _expects_foo(x: Int)
	do
		if x == 0 then return
		super(x)
	end

	redef fun _contract_foo(x: Int)
	do
		_expects_foo
		super(x)
	end

end
```

#### Ensure
```nit
class SubMyClass
	super MyClass
	
	redef fun foo(x: Int): Bool
	is
		ensures(result == true, x > 0)
	do
		[...]
	end
	
	redef fun _ensures_foo(x: Int, result: Bool)
	do
		super
		assert super(x, result)
	end

	redef fun _contract_foo(x: Int, result: Bool): Bool
	do
		var result = super
		_ensures_foo(x, result)
		return result
	end
end
```

Summary

| Annotation    |  Inheritance condition type  | 
| ------------- | -------------| 
| expects       |        and  |
| ensures       |        or  |

## Invocation 

All calls to contract methods are intercepted to call the contract version. For the moment the contracts are systematically called, whether it's an internal or external call. Only calls to super do not execute the contracts.

This part is subject to evolution with a parameter to activate or not the internal call verification.

## Building

Since contracts are expensive in resource, several levels of use have been defined based on the needed verification:

- No contract: all contracts are disable (option `--no-contract`).

- Default: By  default  the  contracts  can  be  defined as "semi-global". I.E. All contracts (ensures, expects)used in the main package are enabled, the expects contracts are enabled (ensures contracts are disable) in direct imported package. Other indirected imported package doesn't have active contract.

- Full contract: Contracts are enable (ensures, expects) on all classes (option `--full-contract`).

## Future work

This pr has been simplified to facilitate the reviewing.

Future features: 

- Support class invariant contracts 
- Ability to put contracts on auto-getters and setter
- Detection of query or command to avoid the contract to change the object state (this functionality will probably be an advice because it's difficult to define if a method has no side effect.)
- `Old` support for the ensures contract  to refer at the same value during the expects condition

Some pr will follow to use the contacts directly in the lib (collection...).

### Note 

Currently the syntax of contracts is strict, only comparisons and method calls returning bouleans can be used. Indeed the arguments passed to the annotation of contract are directly used for assert. A solution would be given a more open syntax for more complex contracts instead of using a method. But this would cause problems in the syntax of annotations. If you have any suggestion don't hesitate. An issue is available to talk about the potential syntax. (#1340)

### CODE REVIEW

File: src/contracts.nit [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+				return drived_callsite
+			end
+		end
+		return callsite
+	end
+end
+
+redef class ANode
+	private fun create_contracts(v: ContractsVisitor)do end
+	private fun check_callsite(v: CallSiteVisitor)do end
```
Comment (privat) [inline src/contracts.nit]: space before `do` here (and above)

File: src/contracts.nit [hunk 2]
```diff
@@ -0,0 +1,885 @@
...
+	var visited_method: APropdef is noinit
+
+	redef fun visit(node)
+	do
+		node.check_callsite(self)
+		node.visit_all(self)
+	end
+
+	# Check if the callsite calls a method with a contract.
+	# It's the case the callsite is replaced by another callsite to the contract method.
```
Comment (privat) [inline src/contracts.nit]: "**If** it's the case," ?

File: src/contracts.nit [hunk 3]
```diff
@@ -0,0 +1,885 @@
...
+	private fun check_callsite(v: CallSiteVisitor)do end
+end
+
+redef class AAnnotation
+
+	# Returns the conditions of annotation parameters in the form of and expr
+	private fun construct_condition(v : ContractsVisitor): AExpr
+	do
+		var first_arg = n_args.first
+		n_args.remove_at(0)
```
Comment (privat) [inline src/contracts.nit]: what are `n_args` and what is `first_arg` here?

File: src/contracts.nit [hunk 4]
```diff
@@ -0,0 +1,885 @@
...
+		# Create the expected condition recursively with the annotation parameters
+		var n_condition = self.construct_and_expr(v.ast_builder,n_args, first_arg)
+		return n_condition
+	end
+
+	# Method that recursively traverses the argument list of the annotation
+	# FIXME: Do something more clear
+	private fun construct_and_expr(ast_builder : ASTBuilder, args:  ANodes[AExpr], n_expr: AExpr): AExpr
```
Comment (privat) [inline src/contracts.nit]: I don't understand what is the point of the method.

File: src/contracts.nit [hunk 5]
```diff
@@ -0,0 +1,885 @@
...
+	do
+		contract.adapt_block_to_contract(v, n_mpropdef)
+		mpropdef.has_contract = true
+		n_mpropdef.do_all(v.toolcontext)
+	end
+end
+
+redef class ASelfExpr
+	# Defined the self type
+	fun set_mtype(mtype: nullable MType)
```
Comment (privat) [inline src/contracts.nit]: this could be in `AExpr` in `typing`, no?

File: src/contracts.nit [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+		var rtype = return_mtype
+		var msignature = adapt_to_condition
+		if rtype != null then
+			msignature.mparameters.add(new MParameter("result", rtype, false))
+		end
+		return msignature
+	end
+
+	# Return a partial copy of self
+	private fun copy: MSignature
```
Comment (privat) [inline src/contracts.nit]: clone?

File: src/contracts.nit [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+	end
+end
+
+redef class MSignature
+	# Adapt signature for a expect condition
+	# Removed the return type is it not necessary
+	private fun adapt_to_condition(return_mtype: nullable MType): MSignature do
+		var adapt_msignature = self.copy
+		adapt_msignature.set_return(return_mtype)
+		return adapt_msignature
```
Comment (privat) [inline src/contracts.nit]: just do it, no need ad hoc helpers:

`return new MSignature(self.mparameters.to_a, return_mtype)`

File: src/contracts.nit [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+		return msignature
+	end
+
+	# Return a partial copy of self
+	private fun copy: MSignature
+	do
+		return new MSignature(mparameters.to_a, return_mtype)
+	end
+
+	private fun set_return(return_mtype: nullable MType)
```
Comment (privat) [inline src/contracts.nit]: not a fan of `set_return` but I can understand as `mparameters` is mutable

File: src/contracts.nit [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+	do
+		self.return_mtype = return_mtype
+	end
+end
+
+redef class ASignature
+	# Create an array of AExpr representing the call of every parameters
```
Comment (privat) [inline src/contracts.nit]: `call` is not a good term as it is used for function calls (and loosely for message sendings)

File: src/contracts.nit [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+	end
+end
+
+# This visitor checks the `AMethPropdef` and the `AClassDef` to check if they have a contract annotation or it's a redefinition with a inheritance contract
+private class ContractsVisitor
+	super Visitor
+
+	var toolcontext: ToolContext
+
+	# The mainmodule is used to define if the finded contract need to be build
```
[the remark this answers is not in the document]
Comment (Delja) [reply src/contracts.nit]: The main module is stored to check if the contract must be generated or not (depending on the selected policy no-contract, full-contract or default).

File: src/contracts.nit [hunk 11]
```diff
@@ -0,0 +1,885 @@
...
+	do
+		node.check_callsite(self)
+		node.visit_all(self)
+	end
+
+	# Check if the callsite calls a method with a contract.
+	# It's the case the callsite is replaced by another callsite to the contract method.
+	private fun drive_method_contract(callsite: CallSite): CallSite
+	do
+		if callsite.mpropdef.has_contract and not callsite.recv.need_anchor then
```
[the remark this answers is not in the document]
Comment (privat) [reply src/contracts.nit]: You didn't remove the `not callsite.recv.need_anchor`

File: src/contracts.nit [hunk 12]
```diff
@@ -0,0 +1,847 @@
...
+		node.visit_all(self)
+	end
+
+	# Check if the callsite calls a method with a contract.
+	# If it's the case the callsite is replaced by another callsite to the contract method.
+	private fun drive_method_contract(callsite: CallSite): CallSite
+	do
+		if callsite.mpropdef.has_contract and not callsite.recv.need_anchor then
+			var recv = callsite.recv
+			if recv.need_anchor then recv = callsite.anchor.as(not null)
```
Comment (privat) [inline src/contracts.nit]: not really.

anchor is the resolved context to solve a type with formal type.

eg

```nit
class Foo[E: Bar]
   fun foo do
      var x = new Array[E]
      x.add(new Bar)
      var b: Bar = x.pop # `pop` have contracts
   end
end
```

here the static type of `x` is `Array[E]` (and the static type of `self` is `Foo[E]` but it's not important here).
The problem is that `E` is a formal type, so how to reason with it, for instance: "is `var b: Bar = x.pop` valid?"
The anchor is used to reason about them. In the method, the anchor is `Foo[Bar]`, it is used to know how formal types are statically bounded.
The anchor is needed as a local context (as the mmodule) to answer some type-related questions.

File: src/contracts.nit [hunk 13]
```diff
@@ -0,0 +1,848 @@
...
+		return false
+	end
+
+	# Display warning if no contract is defined at introduction `expect`,
+	# because if no contract is defined at the introduction the added
+	# contracts will not cause any error even if they are not satisfied.
+	#
+	# exemple
+	#
+	# class A
```
Comment (privat) [inline src/contracts.nit]: use `~~~nitish` or else the doc rendering will be a mess

**Pull request merged**